<a href="https://colab.research.google.com/github/changsksu/KState_IMSE841/blob/main/Concept_of_Contrast.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Concept of Contrast of 2^k factorial design
Example from Myers 4ed section 3.2

In [2]:
import math
import statistics
import numpy as np
import scipy.stats
from scipy import stats
from scipy.stats import f_oneway
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
from statsmodels.formula.api import ols
import statsmodels.stats.multicomp as mc

In [3]:
# the Y data is the sum of all four observation
# contrasts
x1=np.array((-1,+1,-1,+1))
Y=np.array([580,617,553,597])
contrastA=np.matmul(Y, x1)/(2*len(Y))
contrastA

10.125

By changing the A factor at -1 level (15%) to +1 level (25%), the vescosity increases 10.125 units.

In [ ]:
# contrasts
Y=np.array([580,617,553,597])
contrastB=np.matmul(Y, x2)/(2*len(Y))
contrastB

-5.875

In [8]:
# make sure that you use the raw format in Github
# the data is under Dr. Chang's github changsksu folder KState_IMSE541
# use the URL once you find or store the data file (use the raw tab)
# the following input statement is needed to use pandas dataframe feature
import pandas as pd
data = pd.read_csv('https://raw.githubusercontent.com/changsksu/KState_IMSE841/refs/heads/main/Myers%20Table%203.1.csv', sep=',', na_values=".")
# the foloowing statement only retrieves the first 5 obs
data.head()

A   B  Viscosity
0  0.15  20        145
1  0.15  20        148
2  0.15  20        147
3  0.15  20        140
4  0.25  20        158

In [4]:
# click on Choose File to read in a data file from you local machine
from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving Myers Table 3.1.csv to Myers Table 3.1.csv
User uploaded file "Myers Table 3.1.csv" with length 226 bytes


In [6]:
# pandas provides a better view of structured data with headings
# data is assigned to the pandas object called data1
# xxxx.csv is the file name that was inputed in the previsous code block
import pandas as pd
data1 = pd.read_csv('Myers Table 3.1.csv', sep=',', na_values=".")
data1.head()

A   B  Viscosity
0  0.15  20        145
1  0.15  20        148
2  0.15  20        147
3  0.15  20        140
4  0.25  20        158

In [9]:

# first create a column of 1s but it is list
# conver the list into an array
# Example 4.13 pp 161 Montgomery 7ed
one=np.array([1]*16)
one


array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [24]:
x1=np.array([-1,-1,-1,-1,1,1,1,1,-1,-1,-1,-1,1,1,1,1])
x2=np.array([-1,-1,-1,-1,-1,-1,-1,-1,1,1,1,1,1,1,1,1])
Xp=np.asmatrix([one, x1, x2])
Xp

matrix([[ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1],
        [-1, -1, -1, -1,  1,  1,  1,  1, -1, -1, -1, -1,  1,  1,  1,  1],
        [-1, -1, -1, -1, -1, -1, -1, -1,  1,  1,  1,  1,  1,  1,  1,  1]])

In [25]:
X=Xp.transpose()
X

matrix([[ 1, -1, -1],
        [ 1, -1, -1],
        [ 1, -1, -1],
        [ 1, -1, -1],
        [ 1,  1, -1],
        [ 1,  1, -1],
        [ 1,  1, -1],
        [ 1,  1, -1],
        [ 1, -1,  1],
        [ 1, -1,  1],
        [ 1, -1,  1],
        [ 1, -1,  1],
        [ 1,  1,  1],
        [ 1,  1,  1],
        [ 1,  1,  1],
        [ 1,  1,  1]])

In [27]:
# Contrast A
XpX= np.matmul(Xp, X)
XpX

matrix([[16,  0,  0],
        [ 0, 16,  0],
        [ 0,  0, 16]])

In [29]:
Y=np.array(data['Viscosity'])
Y=Y.transpose()
XpY= np.matmul(Xp, Y)
XpY

matrix([[2347,   81,  -47]])

In [30]:
# inverse of the X' * X matrix
XpX= np.matmul(Xp, X)
XXinv=np.linalg.inv(XpX)
XXinv

matrix([[0.0625, 0.    , 0.    ],
        [0.    , 0.0625, 0.    ],
        [0.    , 0.    , 0.0625]])

In [31]:
# compute the beta estimation
beta=np.matmul(XXinv, XpY.transpose())
beta

matrix([[146.6875],
        [  5.0625],
        [ -2.9375]])

In [ ]:
beta.shape

(3, 1)

# Now We can use the ols model to compute betas

In [16]:
A_mid = (data['A'].min() + data['A'].max()) / 2
A_range = (data['A'].max() - data['A'].min())
B_mid = (data['B'].min() + data['B'].max()) / 2
B_range = (data['B'].max() - data['B'].min())

# Coded variables x1 and x2
data['x1'] = (data['A'] - A_mid) / (A_range/2)
data['x2'] = (data['B'] - B_mid) / (B_range/2)

# Fit a multiple linear regression model using sklearn's LinearRegression
X = data[['x1', 'x2']]
y = data['Viscosity']

model = ols('y ~ x1 + x2 + x1*x2', data).fit()
print(model.summary())



                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.846
Model:                            OLS   Adj. R-squared:                  0.808
Method:                 Least Squares   F-statistic:                     21.99
Date:                Fri, 27 Sep 2024   Prob (F-statistic):           3.63e-05
Time:                        20:03:56   Log-Likelihood:                -37.384
No. Observations:                  16   AIC:                             82.77
Df Residuals:                      12   BIC:                             85.86
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    146.6875      0.723    203.003      0.0

/usr/local/lib/python3.10/dist-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  res = hypotest_fun_out(*samples, **kwds)


Note that the main effect = 2 * beta